core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -1,187 +1,337 @@
+            "Ring buffer should be exactly the intended size"
+        );
     }
 
-    fn report_cluster_slots_size(&self) {
-        if self.last_report.should_update(10_000) {
-            let (cluster_slots_cap, pubkeys_capacity) = {
-                let cluster_slots = self.cluster_slots.read().unwrap();
-                let cluster_slots_cap = cluster_slots.len();
-                let pubkeys_capacity = cluster_slots
-                    .iter()
-                    .map(|(_slot, slot_pubkeys)| slot_pubkeys.read().unwrap().capacity())
-                    .sum::<usize>();
-                (cluster_slots_cap, pubkeys_capacity)
-            };
-            let (validator_stakes_cap, validator_pubkeys_len) = {
-                let validator_stakes = self.validator_stakes.read().unwrap();
-                let validator_len = validator_stakes
-                    .iter()
-                    .map(|(_pubkey, vote_accounts)| vote_accounts.vote_accounts.capacity())
-                    .sum::<usize>();
-                (validator_stakes.capacity(), validator_len)
+    fn update_internal(
+        &self,
+        root: Slot,
+        validator_stakes: &HashMap<Pubkey, u64>,
+        epoch_slots_list: Vec<EpochSlots>,
+    ) {
+        let total_stake = self.total_stake.load(std::sync::atomic::Ordering::Relaxed);
+        // Prepare a range of slots we will store in the datastructure
+        // We will discard slots at or before current root or too far ahead.
+        let slot_range = (root + 1)..root.saturating_add(CLUSTER_SLOTS_TRIM_SIZE as u64 + 1);
+        // ensure the datastructure has the correct window in scope
+        self.roll_cluster_slots(validator_stakes, &slot_range);
+
+        let cluster_slots = self.cluster_slots.read().unwrap();
+        for epoch_slots in epoch_slots_list {
+            //filter out unstaked nodes
+            let Some(&sender_stake) = validator_stakes.get(&epoch_slots.from) else {
+                continue;
             };
+            let updates = epoch_slots
+                .to_slots(root)
+                .filter(|slot| slot_range.contains(slot));
+            // figure out which entries would get updated by the new message and cache them
+            for slot in updates {
+                let RowContent {
+                    slot: s,
+                    total_support,
+                    supporters: map,
+                } = Self::get_row_for_slot(slot, &cluster_slots).unwrap();
+                debug_assert_eq!(*s, slot, "Fetched slot does not match expected value!");
+                let slot_weight_f64 =
+                    total_support.load(std::sync::atomic::Ordering::Relaxed) as f64;
+                // once slot is confirmed beyond `DUPLICATE_THRESHOLD` we should not need
+                // to update the stakes for it anymore
+                // floats can be funny, give us a bit of margin
+                if slot_weight_f64 / total_stake as f64 > FREEZE_THRESHOLD * 1.1 {
```

Comment:
The actual lockout logic

Reply:
No this needs to be about 80-90% to not starve repair from peers. As long as it is below 90% we get most of the performance benefits.